File: lua/p4.lua
#!/usr/bin/lua

function TableLength(T)
  local count = 0
  for _ in pairs(T) do count = count + 1 end
  return count
end

function ReverseInteger(val)
  local reversed = 0
  while (val > 0) do
    reversed = (10 * reversed) + (val % 10)
    val = math.floor(val / 10)
  end
  return reversed
end

function IsPalindrome(n)
  return n == ReverseInteger(n)
end

count = 3
if (TableLength(arg) > 2) then
  count = arg[1]
end
tmp = count
max = 0
while (tmp > 0) do
  max = (max * 10) + 9
  tmp = tmp - 1
end
a = max
largest = 0
while (a >= (10 * count)) do
  b = 0
  db = 0
  if (a % 11 == 0) then
    b = max
    db = 1
  else
    b = 11 * (max / 11)
    db = 11
  end
  while (b >= a) do
    if ((a * b) <= largest) then
      break
    elseif (IsPalindrome(a * b))then
      largest = a * b
    end
    b = b - db
  end
  a = a - 1
end
print(string.format("Largest palindrome multiples of %i digits: %i", count, largest))

File: perl/p4.pl
#!/usr/bin/perl

sub ReverseInteger {
  $val = int($_[0]);
  $reversed = 0;
  while ($val > 0){
    $reversed = ($reversed * 10) + ($val % 10);
    $val = int($val / 10);
  }
  return $reversed;
}

sub IsPalindrome {
  return $_[0] == ReverseInteger($_[0]);
}

$count = 3;
if (scalar(@ARGV) > 0){
  $count=$ARGV[0];
}
$tmp = $count;
$max=0;
while($tmp > 0){
  $max = ($max * 10) + 9;
  $tmp -= 1;
}
$a = $max;
$largest = 0;
while($a >= (10 * $count)){
  $b = 0;
  $db = 0;
  if($a % 11 == 0){
    $b = $max;
    $db = 1;
  }else{
    $b = 11 * ($max / 11);
    $db  = 11;
  }
  while($b >= $a){
    if($a * $b <= $largest){
      break;
    }elsif(IsPalindrome($a * $b)){
      $largest = $a * $b;
    } 
    $b -= $db;
  }
  $a -= 1;
}

printf $largest;
